idunn/api/places_list.py

```diff
@@ -228,33 +190,61 @@ async def get_places_bbox_impl(
-        )
-        return [BragiPOI(f) for f in bragi_response.get("features", [])]
+async def select_datasource(params):
+    if pj_source.bbox_is_covered(params.bbox):
+        await select_datasource_for_france(params)
+    else:
+        await select_datasource_outside_france(params)
 
-    # Default source (OSM) with category or class/subclass filters
-    if params.raw_filter:
-        filters = [MimirPoiFilter.from_url_raw_filter(f) for f in params.raw_filter]
+
+async def select_datasource_for_france(params):
+    if any(cat in params.category for cat in TRIPADVISOR_CATEGORIES_COVERED_IN_FRANCE):
+        params.source = PoiSource.TRIPADVISOR
+    elif await is_brand_detected_or_pj_category_cover(params):
+        params.source = PoiSource.PAGESJAUNES
     else:
-        filters = [f for c in params.category for f in c.raw_filters()]
-
-    bbox_places = await run_in_threadpool(
-        fetch_es_pois,
-        poi_es_index_name,
-        filters=filters,
-        bbox=params.bbox,
-        max_size=params.size,
-    )
-    return [POI(p["_source"]) for p in bbox_places]
+        params.source = PoiSource.OSM
+
+
+async def select_datasource_outside_france(params):
+    if any(cat in params.category for cat in TRIPADVISOR_CATEGORIES_COVERED_WORLDWIDE):
+        params.source = PoiSource.TRIPADVISOR
+    else:
+        params.source = PoiSource.OSM
+
+
+async def is_brand_detected_or_pj_category_cover(params):
+    return params.q or (params.category and all(c.pj_what() for c in params.category))
+
+
+async def _fetch_extended_bbox(bbox_extended, params, places_list):
+    original_bbox = params.bbox
+    original_bbox_width = original_bbox[2] - original_bbox[0]
+    original_bbox_height = original_bbox[3] - original_bbox[1]
+    original_bbox_size = max(original_bbox_height, original_bbox_width)
+    if original_bbox_size < EXTENDED_BBOX_MAX_SIZE:
+        # Compute extended bbox and fetch results a second time
+        scale_factor = EXTENDED_BBOX_MAX_SIZE / original_bbox_size
+        new_box = scale(box(*original_bbox), xfact=scale_factor, yfact=scale_factor)
+        params.bbox = new_box.bounds
+        bbox_extended = True
+        places_list = await _fetch_places_list(params)
+    return bbox_extended, places_list
+
+
+async def _fetch_places_list(params: PlacesQueryParam):
+    datasource = DatasourceFactory().get_datasource(params.source)
+    return await datasource.get_places_bbox(params)
+
+
+class DatasourceFactory:
+    def get_datasource(self, source_type: PoiSource) -> Datasource:
+        """Get the matching datasource to fetch POIs"""
+        if source_type == PoiSource.TRIPADVISOR:
+            return Tripadvisor()
+        if source_type == PoiSource.PAGESJAUNES:
+            return PagesJaunes()
+        if source_type == PoiSource.OSM:
+            return Osm()
+        logger.error("%s is not a valid source type, OSM source is used as fallback", source_type)
+        return Osm()
```

Comment:
I think we should crash the function as this could only happen if there is a critical issue in our implementation.

I would reserve this kind of behavior (error log + fallback) for cases where an external API returns an unexpected result.

Reply:
Definitely